### DISCUSSION TITLE
Dealing with sources of randomness in CV-based grid search

### DISCUSSION THREAD

**Discussion by fkiraly** [alan-turing-institute/sktime#5735] [Q&A]
Answering a question by @jwhyte from discord here due to length.
Original questions:

---

Seeking guidance on sktime/sklearn model fitting for best results: 
I've written my own grid search routine for a classifier pipeline (say "clf") given my concerns about results/reproducibility in a bug report (https://github.com/sktime/sktime/issues/5702). 

This routine seems to work quite well doing k-fold cross-validation (CV) on a training set (a test set was excluded previously) whilst scanning across the grid of clf's classifier parameters, setting random_state deterministically for each grid point. The mean of k cv scores is evaluated for each grid point, and this mean is used to show the best parameter value found in the grid. 

We can then take those best parameters and fit the classifier, with a particular random_state assigned, on the entire Xtrain, ytrain. We can then use clf to predict labels for the test set's Xtest, and get the score (classification accuracy). 

The complication I've found is that using 10 different values for  random_state in the clf fitting  gave a range of scores (difference between best and worst results) of about 3.8%. That is, this part of the entire fitting/estimation process might be much more influential than (e.g.,) the differences in the third decimal place seen for the best few values found from the CV/tuning grid search. 

Questions: 
does this sensitivity to random_state suggest I should be using more than one random_state per grid point when I've doing cross_validation?

Assuming the cross-validation has produced a valid result for classifier parameters:
Is it good practice to fit my model to Xtrain, ytrain for a reasonable (more than 10?) number of random_state values and keep the trained model with the best performance?
 
Extra context: thinking about classical non-linear optimisation problems: we cannot guarantee that we will find the optimum value of our objective function. So, to increase our chances of getting a result near the optimum, we run the optimisation with a large number of random initial conditions, expecting that this will give us a better chance to explore the parameter space. Is this kind of thinking appropriate for my concerns about how many values of random_state are needed for a model fitting process in sktime?

**Comment by fkiraly** [comment]
I'll separate the question into "sktime features" and "methods" parts. The methodology part touches some interesting research questions on evaluation of algorithms.

sktime features - doing repeats for cv, switching random state

methodology - how to evaluate algorithms or tune them

**Comment by fkiraly** [comment]
sktime features:

* I notice there is currently no way to do repeats in the searches based on CV. I think the best way to support this would be a `Repeat` compositor for splitters imo.
* I also notice that the grid search currently does not handle the random seed properly when it spawns copies - it is an interesting design question how to resolve this.

**Comment by fkiraly** [comment]
methodology:

I think your points are quite interesting and they touch multiple questions about evaluation and tuning.

At the start, I would like to emphasize that it is important to distinguish:

* measures that ensure fair/accurate evaluation of algorithms
* effective tuning algorithms

There is an interaction between the two, but the one does not necessarily imply, or require the other.

Further, there are also three main sources of stochastic variation in performace estimates:

* the data
* the evaluation scheme, e.g., re-sampling - choice of which data is training or test
* randomness in the algorithm

### on effective tuning

I think this is what your question is about - should you use repetitions to enable more effective tuning?
One could imagine that averaging gives you more accuate performance estimates.

However, this impression is partially misleading, as you repeat a draw only from one of the sources of stochastic variation - the randomness of the algorithm. Since the data sub-sample in this case will not change, the increase in accuracy is misleading, as you are not impacting the variance from the data (it is constant).

Now, empirically, it still could help to repeat evaluation in grid search - but whether it does needs to be checked in a benchmarking experiment where you treat the grid search as part of an estimator and evaluate it like you would any other algorithm of the same type.

In my informal (not published etc) experience, repeating evaluations does not tend to make an impact in grid search, as there are often very similar values that give you very similar performance, so if one of these is pushed up by the noise, another equivalent value will be pushed own.

### on evaluation

For evaluation, it is important to take all sources into account, and which ones you do in your setup will give a different statement about the algorithm.

The most notable difference is whether you want it about:

1. the fitted algorithm, i.e., without refitting - if you now go use it on unseen data, similar of the kind you are dealing with
2. the algorithm in general, including fitting

Somewhat counterintuitively, you can obtain a robust statement on the fitted algorithm (option 1) via a single split, you do not cross-validate. You just fit it once, and then you use x% of test data held out from developing the algorithm.

Getting a statement about 2 can be done in two ways:

* on a representative selection of data sets considered i.i.d. - then it is again enough to run single splits many times. If you cross-validate or repeat, that is ok, though you need to be careful with significances.
* if you only have a single dataset, then it is a partly open research question how to obtain reliable guarantees. It is no longer unsolved, ut it has been for a while, and results are not well-known and counterintuitive. For instance, robust estimates may use cross-validation where the test set is larger than the training set, and there are some negative results for going the other way. This area touches the theory of symmetric probability distributions and is a true methodological rabbit hole. Let me know if you are interested in details.

**Comment by jmwhyte** [comment]
Thanks @fkiraly for a thorough reply. It's particularly good that you've set out the important concepts under "....three main sources of stochastic variation in performace estimates:" which formalises some of the intuition I had, and gives us a standard language for discussion. 

I can appreciate that if you're averaging cross-validation (cv) scores for a particular `random_state`, that just doing more of this with a different `random_state` in the grid search might not create much of a different result. (However, for other ways of scoring, perhaps this would be a different result?) 

I am certainly interested in learning more about this universe of concepts, but time is against me, I have only one data set, and my results so far (using a few different classifiers and transformers on panel data) are struggling to get from about 65% accuracy up to an acceptable minimum closer to 75%. I need to be sure that I don't miss potential improvements to accuracy that are hidden (e.g. by performing only one fit of a trained model with parameters found through the cv process), as this will surely mean I do not choose the best classifier pipeline. I guess this relates to your phrase "measures that ensure fair/accurate evaluation of algorithms".

So returning to my questions:
"does this sensitivity to random_state suggest I should be using more than one random_state per grid point when I've doing cross_validation?" - it seems the answer is NO, which is good, as then we can proceed to the next question...

"Is it good practice to fit my model to Xtrain, ytrain for a reasonable (more than 10?) number of random_state values and keep the trained model with the best performance?"

We can see now that as the train-test split is fixed earlier, of your list of sources of stochastic variation, the question is about whether it is useful/appropriate/recommended to explore only the effect of "randomness in the algorithm" by fitting the "best" classifier for a number of `random_states`? 

I can appreciate that this is not a thorough evaluation of a classifier fitting/prediction process, but it still might help find better results than a single fit to Xtrain,ytrain can. Of course, we know (from the above content) that results are conditional on the other sources of variation being ignored by the fitting stage. It is good to be explicit about this!

Anyway, thanks for all the advice, and I look forward to the continuing evolution of the sktime docs to give us rookies further insights into good practice in applications.

**Reply by fkiraly** [reply]
> Is it good practice to fit my model to Xtrain, ytrain for a reasonable (more than 10?) number of random_state values and keep the trained model with the best performance?

That seems like an unusual thing to do, but it does not necessarily mean it is a problem.

What is important if you do sth like this is that you have an untouched test set for evaluation at the end - the more things like this you do, the more you do "manual overfitting", so you need a means to evaluate the impact of that.

**Reply by jmwhyte** [reply]
Yes, agreed. I think that (going back to my non-linear optimisation example) these questions are somewhat motivated by not having much knowledge about how the TSC are fit + seeing such a big difference in scores between fits. Maybe this means that, for at least some of the TSC, that the "fit" method is not trying out a reasonable number of possible parameters, which is why manually setting `random_state` and exploring that seems to produce better results. 
I recently found this article 
https://towardsdatascience.com/manipulating-machine-learning-results-with-random-state-2a6f49b31081

and a key point from a first reading are:
* "2) Use the prediction results as an interval

Since the results vary across a range, you can opt to report the cross validation results as a range. Repeat the run with different seeds in order to produce a confidence interval that you can report with. Its the range that one can comfortably say the performance band of the model really lies within."

Reporting a range of accuracy scores from `clf.predict()` seems as though it might lead to a more robust evaluation process for a suite of models. However, having chosen which classifier to keep, which fitted classifier of this class do you use? Again, when `random_state` seems very influential in scores, this does seem to bring in a selection issue. 

The article mentions the influence of `random_state`:
" If the accuracy results vary wildly with the seed, it probably means that the model is not robust and you should consider improving your methods to better fit the data."
(It's not a trivial thing for novices to improve quite complex code, so "improving your methods" might not be achievable in the short term... Maybe the best we can do is ask for more details about how model fitting happens?)

Ok, so we've learned that "manual overfitting" should be avoided. I did know about the importance of having training/test data, but I'd be happy to have recommendations on other pitfalls of TSC.

**Reply by jmwhyte** [reply]
I'll just quickly add some further explanation on why I'm interested in this topic. This morning I took the classifier hyper-parameters found from a grid search with 5-fold cv, and for a range of random states, fit this to the entire training set, recording the accuracy scores found predicting labels for the test set. Between best and worst scores is (in absolute terms) a difference of 9.16%. Clearly this kind of variability is going to make evaluation of fits across HP very difficult! I think it would be extremely interesting to see what level of variability is seen for sktime/sklearn classifiers on some suitable test data.
